### ISSUE TITLE
Amplify v6: It is not possible to set boundaries in REST requests with multipart

### ISSUE THREAD

**Issue by kingferiol** [aws-amplify/amplify-js#12638]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-js/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### JavaScript Framework

React

### Amplify APIs

REST API

### Amplify Categories

api

### Environment information

<details>

```
# Put output below this line

  "dependencies": {
    ...,
    "@aws-amplify/api-rest": "^4.0.5",
    "@aws-amplify/auth": "^6.0.5",
    "@aws-amplify/ui-react": "^6.0.3",
  },

```

</details>

### Describe the bug

When i try to submit a Rest request that contains Multipart the framework sets Content-Type to multipart/form-data and it prevents the creatorin of boundary 

Content-Type: multipart/form-data; boundary=---------------------------3513053024223207799458523308

https://github.com/aws-amplify/amplify-js/blob/a353e0d777a1ca5e57aa4b2cda48d147ee1d4803/packages/api-rest/src/apis/common/handler.ts#L54

This is a change compared to the v.5.X version.

Am i missing something? Is there any documentation about this change?

### Expected behavior

Implement the same behavior that we had with the previous version

### Reproduction steps

  const form = new FormData();
  form.append(parameterName, new Blob([file], { type: file.type }), file.name);
  post({ apiName, path, options: { body: form } }).response.body.json();

### Code Snippet

```javascript
// Put your code below this line.

```

### Log output

<details>

```
// Put your logs below this line


```

</details>

**Comment by kingferiol** [comment]
Am I missing something for file upload? Any workaround? Should I revert the v6 upgrade?

**Comment by chrisbonifacio** [comment]
Hi @kingferiol thanks for raising this issue. 
Can you explain your use case for why you are using `post` to upload a file rather than the `storage` library?

Also, is this code snippet the same one that was working in v5?
>const form = new FormData();
form.append(parameterName, new Blob([file], { type: file.type }), file.name);
post({ apiName, path, options: { body: form } }).response.body.json();

Were you previously setting the `Content-Type` header with a boundary before? The code snippet above is only setting a body in the options.

**Comment by kingferiol** [comment]
Hi @chrisbonifacio, thanks for the answer

We are trying to upload a file with REST API. Based on the documentation the `store` function is used just with S3 bucket 

With V5 we were able to achieve the result with the following code 

`
const body = new FormData();
body.append(parameterName, file);
API.post(apiName, path, { body })
`

> Content-Type: multipart/form-data; boundary=----WebKitFormBoundaryAC82h7vSE6v2d07n

 The browser was evaluating for us the Content-Type

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type

Best

Maurzio

**Comment by dvalbuena1** [comment]
Same issue here.
When trying to upload a file via API REST, looks like V6 is not setting the boundary as mentioned by @kingferiol 

In fact, when comparing the requests through Dev tools the boundary is missing in V6:
V5:
![image](https://github.com/aws-amplify/amplify-js/assets/44622851/888b4505-3bc9-4161-8d06-d5934a5d5547)

V6:
![image](https://github.com/aws-amplify/amplify-js/assets/44622851/7689d7ec-49c1-4f78-9e45-3317c3e13e0b)

**Comment by svidgen** [comment]
@kingferiol I'm having trouble getting to the bottom of this one.

Can you tell me what authorization rules are you using on the API? And what specific version of 5.x version you were using where this was working correctly?

cc @AllanZhengYP

**Comment by kingferiol** [comment]
HI @svidgen 

With the following configuration, this was working

    "@aws-amplify/api-rest": "^3.5.6",
    "@aws-amplify/auth": "^5.6.6",
    "@aws-amplify/ui-react": "^5.3.2",

After the following upgrade, it is not possible to upload files with REST API

    "@aws-amplify/api-rest": "4.0.7",
    "@aws-amplify/auth": "6.0.7",
    "@aws-amplify/ui-react": "6.0.6",

Authentication is against AWS Cognito

Best

**Comment by svidgen** [comment]
@kingferiol Do you have the `X-Amz-Content-Sha256` listed in `Access-Control-Allow-Headers` in the API Gateway?

**Comment by kingferiol** [comment]
Hi @svidgen 

No, but I didn't change anything on this during the migration (nothing was changed for the API Gateway).

This weekend, I did the following test with v6. At the end of the execution of this code, I drop the `content-type` attribute in `resolvedHeaders` (just in the case of `FormData`). All is working fine browser can compute the boundary by itself.

```
	const resolvedHeaders: Headers = {
		...normalizeHeaders(headers),
		...(resolvedBody
			? {
					'content-type':
						body instanceof FormData
							? 'multipart/form-data'
							: 'application/json; charset=UTF-8',
			  }
			: {}),
	};

```

**Comment by AllanZhengYP** [comment]
Hi @kingferiol 

Thanks a lot for the investigation! I can reproduce the issue. However, I still get signature error when I unset the `content-type` header and only use the header populated by browser. This is because the API Gateway request validator verifies the request payload checksum, however, the signature calculated on client side does not calculate the payload. However it will work if your API is not protected by IAM.

Can you confirm the API you tested is protected by IAM? For example, you might have disabled it for POST. You can inspect your `backend/rest/{apiName}/cli-inputs.json` file. Normally you have:
```
{
  "version": 1,
  "paths": {
    "/main": {
      "name": "/main",
      "lambdaFunction": "basicrestapimain",
      "permissions": {
        "setting": "protected",
        "auth": [
          "create",
          "read",
          "update",
          "delete"
        ],
        "guest": [
          "read"
        ]
      }
    }
  }
}
```

But you could have set it to
```
...
      "permissions": {
        "setting": "open"
      }
...
```

**Comment by kingferiol** [comment]
Hi @AllanZhengYP 

Our authentication for AWS API Gateway is against AWS Cognito.

Best

Maurizio

**Comment by AllanZhengYP** [comment]
Hi @kingferiol I still cannot reproduce the signature success in v5 in my rudimentary reproduction. But I understood the fix that might unblock your case. I just released a fix to our unstable tag: `aws-amplify: 6.0.9-unstable.58ed2f7.0+58ed2f7`, from PR: https://github.com/aws-amplify/amplify-js/pull/12720

Can you give a try on this tag version and see whether it unblocks you?

**Comment by dvalbuena1** [comment]
Hi @AllanZhengYP 
I just tested `6.0.8` and `6.0.9-unstable.58ed2f7.0` and the `6.0.9` version worked! It is setting the `boundary` correctly.
`6.0.8` version did not.

In the meantime, it is save to use that `6.0.9` version, right?

**Comment by kingferiol** [comment]
Hi @AllanZhengYP 

I can confirm the same as @dvalbuena1 

Best

**Comment by cwomack** [comment]
Closing this issue as it should now be resolved with the release of v6.0.9.
